### PR TITLE
SI-8196 Runtime reflection robustness for STATIC impl details

### PR DESCRIPTION
**Pull request by retronym**
Scala's runtime reflection works in few modes. The primary mode reads
reads out the pickled signatures from ScalaSig annotations, if
avaialable. However, these aren't available for Java-defined classes
(obviously) nor for local Scala-defined classes (less obviously.),
and the Scala `Symbol`s and `Types` must be reconstructed from
the Java generic reflection metadata.

This bug occurs in the last case, and is centered in
`FromJavaClassCompleter`.

In that completer, member fields and methods are given an owner
based on the STATIC modifier. That makes sense for Java defined
classes. I'm not 100% if it makes sense for Scala defined classes;
maybe we should just skip them entirely?

This patch still includes them, but makes the ownership-assignment
more robust in the face of STATIC members emitted by the Scala
compiler backend, such as the cache fields for structural calls.
(It's reflection all the way down!). We might not have a companion
module at all, so before we ended up owning those by `NoSymbol`,
and before too long hit the dreaded NSDHNAO crash.

That crash doesn't exist any more on 2.11 (it is demoted to a
-Xdev warning), but this patch still makes sense on that branch.

This commit makes `followStatic` and `enter` more robust when
finding a suitable owner for static members. I've also factored
out the duplicated logic between the two.

Review by @xeno-by

I think this fix should be targetted at 2.10.5 and 2.11.1

### CODE REVIEW

**Comment by xeno-by** [comment]
LGTM. FromJavaClassCompleter is best-effort anyway.

**Comment by retronym** [comment]
Note that the refactored `followStatic` is also used in Java class completers, where the `orElse` should never happen. I don't think this is a problem, but just fully disclosing the context of the patch.

**Comment by retronym** [comment]
@xeno-by I just added a test case from SI-8195

**Comment by soc** [comment]
Might make sense to also test against `Symbol` and `@enum`, both emit static members. (In fact, I experienced an issue with enums before, in which Scala didn't seem to pick up static members from Scala-defined classes at all, although at compile-time, not in reflection, so that commit is probably only fixing half the issues around STATIC.)

**Comment by retronym** [comment]
@xeno-by just pushed an update that fixes the build error.

@soc thanks for the suggestions. I think its enough to test with one static for now. Calls to STATIC Java members are modelled as non-static calls to the companion object. The backend turns them back into invokestatic if `Symbol#isStaticMember`. So, your `enum` macro needs to enter any STATIC methods into the module class of the companion module before the `pickler` phase to make them visible to Scala code.

**Comment by soc** [comment]
@retronym Macros run in typer, so that should be soon enough, ...

**Comment by retronym** [comment]
Pushed a fix for the binary compat error,

**Comment by retronym** [comment]
@xeno-by Can you please re-review the new commits?

File: bincompat-forward.whitelist.conf [hunk 1]
```diff
@@ -173,6 +173,11 @@ filter {
         {
             matchName="scala.reflect.runtime.SymbolLoaders.isInvalidClassName"
             problemName=MissingMethodProblem
+        },
+        {
+            matchName="scala.reflect.runtime.JavaMirrors#JavaMirror.scala$reflect$runtime$JavaMirrors$JavaMirror$$followStatic"
+            problemName=MissingMethodProblem
```
Comment (xeno-by) [inline bincompat-forward.whitelist.conf:179]: Maybe we could just disable mima for the entire package at once?

**Comment by xeno-by** [comment]
Lgtm

**Pull request merged**